app/src/main/java/com/futurice/freesound/feature/common/ui/adapter/RecyclerViewAdapter.java

```diff
@@ -0,0 +1,104 @@
+package com.futurice.freesound.feature.common.ui.adapter;
+
+import com.futurice.freesound.common.utils.AndroidPreconditions;
+import com.futurice.freesound.feature.common.DisplayableItem;
+import com.futurice.freesound.inject.fragment.FragmentScope;
+
+import android.support.annotation.NonNull;
+import android.support.v7.util.DiffUtil;
+import android.support.v7.widget.RecyclerView;
+import android.view.ViewGroup;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+import javax.inject.Inject;
+
+import io.reactivex.Observable;
+import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.schedulers.Schedulers;
+
+/**
+ * Implementation of {@link android.support.v7.widget.RecyclerView.Adapter} for {@link
+ * DisplayableItem}.
+ */
+@FragmentScope
+public final class RecyclerViewAdapter extends RecyclerView.Adapter {
+
+    @NonNull
+    private final List<DisplayableItem> mItems = new ArrayList<>();
+
+    @NonNull
+    private final IListItemComparator mComparator;
+
+    @NonNull
+    private final IViewHolderFactory mInstantiator;
+
+    @NonNull
+    private final IViewHolderBinder<DisplayableItem> mBinder;
+
+    private Executor mExecutor = Executors.newSingleThreadExecutor();
```

Comment:
The idea was to calculate the difference of the lists in the background, which required reading/writing the items in the adapter in a single background thread to ensure the integrity of the data without actual synchronisation (slow). The problem is that the framework does some calls that require reading of the items in the adapter, getItemCount for instance, and this calls happen in the main thread causing potential race conditions. 
Let's enforce main thread for now since calculating the difference in a background thread is not that easy.